File: src/main/java/com/paragon/api/setting/Bind.java
package com.paragon.api.setting;

import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;

/**
 * @author Wolfsurge 
 * @since 09/06/22
 */
public class Bind {

    // The button index
    private int buttonCode;
    
    // Input device
    private Device device;
    
    // Prevent possible spam
    private boolean alreadyPressed;

    public Bind(int buttonCode, Device device) {
        this.buttonCode = buttonCode;
        this.device = device;
    }

    public boolean isPressed() {
        if (buttonCode == 0) {
            return false;
        }

        // Our bind is pressed
        boolean pressed = Keyboard.isKeyDown(buttonCode) && device.equals(Device.KEYBOARD) || Mouse.isButtonDown(buttonCode) && device.equals(Device.MOUSE);

        if (pressed) {
            // We haven't already pressed the key
            if (!alreadyPressed) {
                this.alreadyPressed = true;
                return true;
            } else {
                return false;
            }
        } else {
            this.alreadyPressed = false;
            return false;
        }
    }

    public enum Device {
        /**
         * A key on the keyboard
         */
        KEYBOARD,

        /**
         * A mouse button
         */
        MOUSE
    }

    /**
     * Gets the button name for the GUI
     * @return The button name
     */
    public String getButtonName() {
        return device.equals(Device.KEYBOARD) ? Keyboard.getKeyName(buttonCode) : Mouse.getButtonName(buttonCode);
    }

    /**
     * Gets the button code
     * @return The button code
     */
    public int getButtonCode() {
        return buttonCode;
    }

    /**
     * Sets the button code
     * @param newCode The new button code
     */
    public void setButtonCode(int newCode) {
        this.buttonCode = newCode;
    }

    /**
     * Gets the input device
     * @return The input device
     */
    public Device getDevice() {
        return device;
    }

    /**
     * Sets the input device
     * @param device The new input device
     */
    public void setDevice(Device device) {
        this.device = device;
    }

}

File: src/main/kotlin/com/paragon/api/setting/Bind.kt
package com.paragon.api.setting

import org.lwjgl.input.Keyboard
import org.lwjgl.input.Mouse

/**
 * @author Surge, SooStrator1136
 */
class Bind(var buttonCode: Int, var device: Device) {

    private var alreadyPressed = false

    fun isPressed(): Boolean {
        if (buttonCode == 0) {
            return false
        }

        return if (Keyboard.isKeyDown(buttonCode) && device == Device.KEYBOARD || Mouse.isButtonDown(buttonCode) && device == Device.MOUSE) { // Our bind is pressed
            // We haven't already pressed the key
            if (!alreadyPressed) {
                alreadyPressed = true
                true
            } else {
                false
            }
        } else {
            alreadyPressed = false
            false
        }
    }

    fun getButtonName(): String = if (device == Device.KEYBOARD) Keyboard.getKeyName(buttonCode) else Mouse.getButtonName(buttonCode)

    enum class Device {
        /**
         * A key on the keyboard
         */
        KEYBOARD,

        /**
         * A mouse button
         */
        MOUSE
    }

}